**Integrantes**

- Juan Diego Correa Noreña
- Yessica Bastidas Villada
- Santiago Bustos Pianda
- Juan Carlos Gómez Ramírez

**TALLER #2 - ¿Cuál es el contexto de la ciudad de Nueva York en términos de criminalidad?**


<p><a name="s3"></a></p>

# **Paquetes importados**


In [1]:
#Importar paquetes
import pandas as pd
import numpy as np
import plotly.express as px
from plotly.subplots import make_subplots
from pandas.io.formats.format import get_format_datetime64_from_values

<p><a name="s3"></a></p>

# **Datos llamados a analizar**


In [2]:
#Importar datos
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
#Cargar base de datos 1 (datos hasta el 2022)
Base1=pd.read_csv('/content/drive/MyDrive/NYPD_Arrests_Data__Historic_2.csv',sep='%')
Base1.head()

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (18,20) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


ARREST_KEY  PD_CD                                 PD_DESC  KY_CD  \
0     9930151  848.0  NY STATE LAWS,UNCLASSIFIED MISDEMEANOR  364.0   
1     9929960  905.0             INTOXICATED DRIVING,ALCOHOL  347.0   
2     9930303  793.0                    WEAPONS POSSESSION 3  118.0   
3     9930099  782.0                WEAPONS, POSSESSION, ETC  236.0   
4     9930251  548.0            GAMBLING, DEVICE, POSSESSION  350.0   

                          OFNS_DESC    LAW_CODE LAW_CAT_CD ARREST_BORO  \
0  OTHER STATE LAWS (NON PENAL LAW)  LOC00000UM          M           M   
1    INTOXICATED & IMPAIRED DRIVING  VTL11920M3          M           K   
2                 DANGEROUS WEAPONS  PL 2650201          F           Q   
3                 DANGEROUS WEAPONS  PL 2650101          M           M   
4                          GAMBLING  PL 2253002          M           Q   

   ARREST_PRECINCT  JURISDICTION_CODE  ... PERP_SEX PERP_RACE X_COORD_CD  \
0               18                0.0  ...        M     BLACK   988950.0   
1               84                0.0  ...        M     BLACK   988156.0   
2              101                0.0  ...        M     BLACK  1046242.0   
3               26                0.0  ...        M     BLACK   996579.0   
4              114                0.0  ...        M     BLACK  1002577.0   

   Y_COORD_CD   Latitude  Longitude  \
0    216678.0  40.761414 -73.983035   
1    189697.0  40.687359 -73.985917   
2    157040.0  40.597506 -73.776773   
3    236020.0  40.814496 -73.955461   
4    221583.0  40.774860 -73.933832   

                                         Lon_Lat ARREST_YEAR ARREST_DAY  \
0   POINT (-73.98303549199994 40.76141449100004)        2006    dia_uno   
1   POINT (-73.98591731099998 40.68735856600006)        2006    dia_uno   
2   POINT (-73.77677316599994 40.59750616900004)        2006    dia_uno   
3  POINT (-73.95546100099993 40.814495755000046)        2006    dia_uno   
4        POINT (-73.933831684 40.77485969700007)        2006    dia_uno   

  ARREST_MONTH  
0          uno  
1          uno  
2          uno  
3          uno  
4          uno  

[5 rows x 21 columns]

In [4]:
#Cargar base de datos 2 (datos del 2022)
Base2=pd.read_csv('/content/drive/MyDrive/NYPD_Arrest_Data__Year_to_Date_2.csv',sep=';')
Base2.head()

ARREST_KEY  PD_CD                         PD_DESC  KY_CD  \
0   238552639  205.0            TRESPASS 2, CRIMINAL  352.0   
1   238682242  397.0  ROBBERY,OPEN AREA UNCLASSIFIED  105.0   
2   238723993  153.0                          RAPE 3  104.0   
3   238821533    NaN                          (null)    NaN   
4   239029586    NaN                          (null)    NaN   

           OFNS_DESC    LAW_CODE LAW_CAT_CD ARREST_BORO  ARREST_PRECINCT  \
0  CRIMINAL TRESPASS  PL 1401501          M           K               75   
1            ROBBERY  PL 160050H          F           K               75   
2               RAPE  PL 1302503          F           K               77   
3             (null)  CPL5700600          9           Q              113   
4             (null)  CPL5700600          9           Q              113   

   JURISDICTION_CODE  ... PERP_SEX       PERP_RACE X_COORD_CD  Y_COORD_CD  \
0                  0  ...        M           BLACK    1015028      185895   
1                  0  ...        M           BLACK    1017119      183909   
2                  0  ...        M           WHITE    1003509      185018   
3                  3  ...        M  WHITE HISPANIC    1046367      186986   
4                  3  ...        M  WHITE HISPANIC    1046367      186986   

    Latitude  Longitude                    New Georeferenced Column  \
0  40.676863 -73.889037                POINT (-73.889037 40.676863)   
1  40.671404 -73.881509                POINT (-73.881509 40.671404)   
2  40.674496 -73.930571  POINT (-73.9305713255961 40.6744956865259)   
3  40.679701 -73.776047            POINT (-73.77604735 40.67970059)   
4  40.679701 -73.776047            POINT (-73.77604735 40.67970059)   

  ARREST_YEAR ARREST_DAY ARREST_MONTH  
0       20 22          3            1  
1       20 22          5            1  
2       20 22   dia_seis            1  
3       20 22   dia_ocho            1  
4       20 22         12            1  

[5 rows x 21 columns]

**Descripción de las variables**
- ARREST_KEY: Identificación de cada arresto.
- ARREST_DATE: Fecha del arresto.
- PD_CD: Código de clasificación interna.
- PD_DESC: Descripción de la clasificación interna correspondiente al código PD.
- KY_CD: Código de clasificación interna de tres dígitos (categoría más general que el código PD).
- OFNS_DESC: Descripción de la clasificación interna correspondiente al código KY (categoría más general que la descripción PD).
- LAW_CODE: Cargos del código de ley correspondientes a la Ley Penal del Estado de Nueva York, VTL y otras diversas leyes locales.
- LAW_CAT_CD: Nivel de ofensa: delito grave, delito menor, violación.
- ARREST_BORO: Municipio de detención. B (Bronx), S (Staten Island), K (Brooklyn), M (Manhattan), Q (Queens).
- ARREST_PRECINCT: Precinto donde ocurrió la detención.
- JURISDICTION_CODE: Jurisdicción responsable de la detención. Los códigos de jurisdicción 0 (Patrulla), 1 (Tránsito) y 2 (Vivienda) representan a la policía de Nueva York, mientras que los códigos 3 y más representan jurisdicciones que no son de la policía de Nueva York.
- GRUPO DE EDAD: Edad del perpetrador dentro de una categoría.
- PERP_SEX: Descripción del sexo del perpetrador.
- PERP_RACE: Descripción de la carrera del perpetrador.
- X_COORD_CD: Coordenada X de bloque medio para el sistema de coordenadas planas del estado de Nueva York, zona de Long Island, NAD 83, unidades pies (FIPS 3104).
- Y_COORD_CD: Coordenada Y de bloque medio para el sistema de coordenadas planas del estado de Nueva York, zona de Long Island, NAD 83, unidades pies (FIPS 3104).
- Latitud: Coordenada de latitud para el Sistema de coordenadas global.
- Longitud: Coordenada de longitud para el Sistema de coordenadas global.
- Lon_Lat: Columna de puntos georreferenciados basada en campos de longitud y latitud.

**Fuentes**
- Base de datos 1: https://data.cityofnewyork.us/Public-Safety/NYPD-Arrests-Data-Historic-/8h9b-rp9u
- Base de datos 2: https://data.cityofnewyork.us/Public-Safety/NYPD-Arrest-Data-Year-to-Date-/uip8-fykc

<p><a name="s3"></a></p>

# **DEPURACIÓN: Homologación de categorías, tipos de datos**



###**Tipos de datos**

In [5]:
#Dimensiones de la base de datos 1
Base1.shape

(5303638, 21)

In [6]:
Base1.dtypes

ARREST_KEY             int64
PD_CD                float64
PD_DESC               object
KY_CD                float64
OFNS_DESC             object
LAW_CODE              object
LAW_CAT_CD            object
ARREST_BORO           object
ARREST_PRECINCT        int64
JURISDICTION_CODE    float64
AGE_GROUP             object
PERP_SEX              object
PERP_RACE             object
X_COORD_CD           float64
Y_COORD_CD           float64
Latitude             float64
Longitude            float64
Lon_Lat               object
ARREST_YEAR           object
ARREST_DAY            object
ARREST_MONTH          object
dtype: object

In [7]:
#Dimensiones de la base de datos 2
Base2.shape

(93238, 21)

In [8]:
Base2.dtypes

ARREST_KEY                    int64
PD_CD                       float64
PD_DESC                      object
KY_CD                       float64
OFNS_DESC                    object
LAW_CODE                     object
LAW_CAT_CD                   object
ARREST_BORO                  object
ARREST_PRECINCT               int64
JURISDICTION_CODE             int64
AGE_GROUP                    object
PERP_SEX                     object
PERP_RACE                    object
X_COORD_CD                    int64
Y_COORD_CD                    int64
Latitude                    float64
Longitude                   float64
New Georeferenced Column     object
ARREST_YEAR                  object
ARREST_DAY                   object
ARREST_MONTH                 object
dtype: object

In [9]:
#De la base de datos 2 (Base2) se cambia el nombre de la columna en la posición 17 (New Georeferenced Column)
#Debido a que en (Base1) esta tiene un nombre diferente (Lon_Lat)

Base2 = Base2.rename(columns={'New Georeferenced Column':'Lon_Lat'})
Base2.head()

ARREST_KEY  PD_CD                         PD_DESC  KY_CD  \
0   238552639  205.0            TRESPASS 2, CRIMINAL  352.0   
1   238682242  397.0  ROBBERY,OPEN AREA UNCLASSIFIED  105.0   
2   238723993  153.0                          RAPE 3  104.0   
3   238821533    NaN                          (null)    NaN   
4   239029586    NaN                          (null)    NaN   

           OFNS_DESC    LAW_CODE LAW_CAT_CD ARREST_BORO  ARREST_PRECINCT  \
0  CRIMINAL TRESPASS  PL 1401501          M           K               75   
1            ROBBERY  PL 160050H          F           K               75   
2               RAPE  PL 1302503          F           K               77   
3             (null)  CPL5700600          9           Q              113   
4             (null)  CPL5700600          9           Q              113   

   JURISDICTION_CODE  ... PERP_SEX       PERP_RACE X_COORD_CD  Y_COORD_CD  \
0                  0  ...        M           BLACK    1015028      185895   
1                  0  ...        M           BLACK    1017119      183909   
2                  0  ...        M           WHITE    1003509      185018   
3                  3  ...        M  WHITE HISPANIC    1046367      186986   
4                  3  ...        M  WHITE HISPANIC    1046367      186986   

    Latitude  Longitude                                     Lon_Lat  \
0  40.676863 -73.889037                POINT (-73.889037 40.676863)   
1  40.671404 -73.881509                POINT (-73.881509 40.671404)   
2  40.674496 -73.930571  POINT (-73.9305713255961 40.6744956865259)   
3  40.679701 -73.776047            POINT (-73.77604735 40.67970059)   
4  40.679701 -73.776047            POINT (-73.77604735 40.67970059)   

  ARREST_YEAR ARREST_DAY ARREST_MONTH  
0       20 22          3            1  
1       20 22          5            1  
2       20 22   dia_seis            1  
3       20 22   dia_ocho            1  
4       20 22         12            1  

[5 rows x 21 columns]

In [10]:
#se concatenaron las 2 bases para tener un mejor manejo de ellas, ya que ahora tienen las mismas 21 columnas
Basedatos=pd.concat([Base1,Base2])
Basedatos.head()

ARREST_KEY  PD_CD                                 PD_DESC  KY_CD  \
0     9930151  848.0  NY STATE LAWS,UNCLASSIFIED MISDEMEANOR  364.0   
1     9929960  905.0             INTOXICATED DRIVING,ALCOHOL  347.0   
2     9930303  793.0                    WEAPONS POSSESSION 3  118.0   
3     9930099  782.0                WEAPONS, POSSESSION, ETC  236.0   
4     9930251  548.0            GAMBLING, DEVICE, POSSESSION  350.0   

                          OFNS_DESC    LAW_CODE LAW_CAT_CD ARREST_BORO  \
0  OTHER STATE LAWS (NON PENAL LAW)  LOC00000UM          M           M   
1    INTOXICATED & IMPAIRED DRIVING  VTL11920M3          M           K   
2                 DANGEROUS WEAPONS  PL 2650201          F           Q   
3                 DANGEROUS WEAPONS  PL 2650101          M           M   
4                          GAMBLING  PL 2253002          M           Q   

   ARREST_PRECINCT  JURISDICTION_CODE  ... PERP_SEX PERP_RACE X_COORD_CD  \
0               18                0.0  ...        M     BLACK   988950.0   
1               84                0.0  ...        M     BLACK   988156.0   
2              101                0.0  ...        M     BLACK  1046242.0   
3               26                0.0  ...        M     BLACK   996579.0   
4              114                0.0  ...        M     BLACK  1002577.0   

   Y_COORD_CD   Latitude  Longitude  \
0    216678.0  40.761414 -73.983035   
1    189697.0  40.687359 -73.985917   
2    157040.0  40.597506 -73.776773   
3    236020.0  40.814496 -73.955461   
4    221583.0  40.774860 -73.933832   

                                         Lon_Lat ARREST_YEAR ARREST_DAY  \
0   POINT (-73.98303549199994 40.76141449100004)        2006    dia_uno   
1   POINT (-73.98591731099998 40.68735856600006)        2006    dia_uno   
2   POINT (-73.77677316599994 40.59750616900004)        2006    dia_uno   
3  POINT (-73.95546100099993 40.814495755000046)        2006    dia_uno   
4        POINT (-73.933831684 40.77485969700007)        2006    dia_uno   

  ARREST_MONTH  
0          uno  
1          uno  
2          uno  
3          uno  
4          uno  

[5 rows x 21 columns]

In [11]:
Basedatos.dtypes

ARREST_KEY             int64
PD_CD                float64
PD_DESC               object
KY_CD                float64
OFNS_DESC             object
LAW_CODE              object
LAW_CAT_CD            object
ARREST_BORO           object
ARREST_PRECINCT        int64
JURISDICTION_CODE    float64
AGE_GROUP             object
PERP_SEX              object
PERP_RACE             object
X_COORD_CD           float64
Y_COORD_CD           float64
Latitude             float64
Longitude            float64
Lon_Lat               object
ARREST_YEAR           object
ARREST_DAY            object
ARREST_MONTH          object
dtype: object

###**Homologación de categorias**

In [12]:
#Se mira que tipo de elementos tiene cada variable para corregirlos
Basedatos['ARREST_DAY'].unique()

array(['dia_uno', 'dia_dos', 'dia_tres', 'dia_cuatro', 'dia_cinco', '6',
       '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17',
       '18', '19', '21', '22', '23', '24', '25', '26', '27', '28', '29',
       '30', '31', '20', '3', '5', 'dia_seis', 'dia_ocho', 'dia_siete',
       'dia_diez', 'dia_nueve', '4', '2', '1'], dtype=object)

In [13]:
#se corrigen los que aparecen mal escritos, llevando todos a numero
Basedatos['ARREST_DAY']=Basedatos['ARREST_DAY'].replace({'dia_uno':'1','dia_dos':'2','dia_tres':'3','dia_cuatro':'4','dia_cinco':'5','dia_seis':'6','dia_siete':'7','dia_ocho':'8','dia_nueve':'9','dia_diez':'10'})

In [14]:
Basedatos['ARREST_DAY'].unique()

array(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',
       '13', '14', '15', '16', '17', '18', '19', '21', '22', '23', '24',
       '25', '26', '27', '28', '29', '30', '31', '20'], dtype=object)

In [15]:
Basedatos['ARREST_MONTH'].unique()

array(['uno', '2', 'tres', '4', 4, 5, 6, '6', 'siete', 'agosto', '9', 9,
       10, 11, '11', 'doce', '10', 2, '5', '1', '3', 'dos', 'cuatro',
       'cinco', 'seis'], dtype=object)

In [16]:
Basedatos['ARREST_MONTH']=Basedatos['ARREST_MONTH'].replace({'uno':'1','dos':'2','tres':'3','cuatro':'4','cinco':'5','seis':'6','siete':'7','agosto':'8','doce':'12',4:'4',5:'5',6:'6',9:'9',10:'10',11:'11',2:'2'})

In [17]:
Basedatos['ARREST_MONTH'].unique()

array(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'],
      dtype=object)

In [18]:
Basedatos['ARREST_YEAR'].unique()

array([2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, ' 2014 ',
       '2 0 1 5', '2 0 1 6', '2 0 1 7', '2 0 1 8', '2 0 1 9', '2 0 2 0',
       '2 0 2 1', '20 22'], dtype=object)

In [19]:
Basedatos['ARREST_YEAR']=Basedatos['ARREST_YEAR'].replace({' 2014 ':'2014','2 0 1 5':'2015','2 0 1 6':'2016','2 0 1 7':'2017','2 0 1 8':'2018','2 0 1 9':'2019','2 0 2 0':'2020','2 0 2 1':'2021','20 22':'2022'})
Basedatos['ARREST_YEAR']=Basedatos['ARREST_YEAR'].replace({2006:'2006', 2007:'2007', 2008:'2008', 2009:'2009', 2010:'2010', 2011:'2011', 2012:'2012', 2013:'2013', 2014:'2014'})

In [20]:
Basedatos['ARREST_YEAR'].unique()

array(['2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021',
       '2022'], dtype=object)

In [21]:
#Se creó una nueva columna con la fecha completa
Basedatos['FECHA']=pd.to_datetime({'year': Basedatos['ARREST_YEAR'],
                                'month': Basedatos['ARREST_MONTH'],
                                'day': Basedatos['ARREST_DAY'] })

In [22]:
Basedatos.dtypes 
#La base de datos tiene 1 variable en formato fecha, 12 variables object, 7 variables float y 2 variables enteras

ARREST_KEY                    int64
PD_CD                       float64
PD_DESC                      object
KY_CD                       float64
OFNS_DESC                    object
LAW_CODE                     object
LAW_CAT_CD                   object
ARREST_BORO                  object
ARREST_PRECINCT               int64
JURISDICTION_CODE           float64
AGE_GROUP                    object
PERP_SEX                     object
PERP_RACE                    object
X_COORD_CD                  float64
Y_COORD_CD                  float64
Latitude                    float64
Longitude                   float64
Lon_Lat                      object
ARREST_YEAR                  object
ARREST_DAY                   object
ARREST_MONTH                 object
FECHA                datetime64[ns]
dtype: object

In [23]:
Basedatos.shape

(5396876, 22)

In [24]:
#Se identifican categorias extrañas para las edades
Basedatos['AGE_GROUP'].unique()

array(['25-44', '18-24', '45-64', '<18', '65+', '945', '932', nan, '921',
       '952', '338', '922', '740', '933', '894', '949', '237', '920',
       '947', '912', 'UNKNOWN', '935', '939', '316', '944', '816', '928',
       '951', '352', '708', '948', '194', '938', '895', '956', '946',
       '943', '940', '323', '965', '339', '934', '320', '929', '330',
       '942', '446', '910', '927', '941', '959', '931', '309', '930',
       '955', '723', '314', '896', '937', '936', '954', '200', '925',
       '212', '924', '926', '640', '231', '914', '909', '923', '812',
       '317', '808', '918', '1922', '1018', '1024', '1948', '1048',
       '2007', '1998', '1953', '1942', '1916', '1937', '1947', '1928',
       '1918', '1042', '2000'], dtype=object)

In [25]:
#Se crea un diccionario para los rangos de edades extraños y se ponen en la categoria 'Desconocido'
lista={'945':'Desconocido','949':'Desconocido','932':'Desconocido','921':'Desconocido','952':'Desconocido','338':'Desconocido',
       '922':'Desconocido','740':'Desconocido','933':'Desconocido','894':'Desconocido','237':'Desconocido','920':'Desconocido',
       '947':'Desconocido','912':'Desconocido','UNKNOWN':'Desconocido','935':'Desconocido','939':'Desconocido','316':'Desconocido',
       '944':'Desconocido','816':'Desconocido','928':'Desconocido','951':'Desconocido','352':'Desconocido','708':'Desconocido',
       '948':'Desconocido','194':'Desconocido','938':'Desconocido','895':'Desconocido','956':'Desconocido','946':'Desconocido',
       '943':'Desconocido','940':'Desconocido','323':'Desconocido','965':'Desconocido','339':'Desconocido','934':'Desconocido',
       '320':'Desconocido','708':'Desconocido','708':'Desconocido','929':'Desconocido','330':'Desconocido','942':'Desconocido',
       '446':'Desconocido','910':'Desconocido','927':'Desconocido','941':'Desconocido','959':'Desconocido','931':'Desconocido',
       '309':'Desconocido','930':'Desconocido','955':'Desconocido','723':'Desconocido','314':'Desconocido','896':'Desconocido',
       '937':'Desconocido','936':'Desconocido','954':'Desconocido','200':'Desconocido','925':'Desconocido','212':'Desconocido',
       '924':'Desconocido','926':'Desconocido','640':'Desconocido','231':'Desconocido','914':'Desconocido','909':'Desconocido',
       '923':'Desconocido','812':'Desconocido','317':'Desconocido','808':'Desconocido','918':'Desconocido','1922':'Desconocido',
       '1018':'Desconocido','1024':'Desconocido','1948':'Desconocido','1048':'Desconocido','2007':'Desconocido','1998':'Desconocido',
       '1953':'Desconocido','1942':'Desconocido','1916':'Desconocido','1937':'Desconocido','1947':'Desconocido','1928':'Desconocido',
       '1918':'Desconocido','1042':'Desconocido','2000':'Desconocido'}

In [26]:
Basedatos['AGE_GROUP']=Basedatos['AGE_GROUP'].replace(lista)

In [27]:
Basedatos['AGE_GROUP'].unique()

array(['25-44', '18-24', '45-64', '<18', '65+', 'Desconocido', nan],
      dtype=object)

In [28]:
Basedatos['LAW_CAT_CD'].value_counts( )

M    3540160
F    1515317
V     293419
I      26612
9        216
Name: LAW_CAT_CD, dtype: int64

In [29]:
#se elimina el 9 de los valores encontrados en la variable LAW_CAT_CD ya que se observa que es un dato que no deberia estar
lista9=Basedatos[Basedatos['LAW_CAT_CD']=='9'].index
Basedatos=Basedatos.drop(lista9)

In [30]:
Basedatos['LAW_CAT_CD'].value_counts( )

M    3540016
F    1515259
V     293407
I      26610
Name: LAW_CAT_CD, dtype: int64

In [31]:
Basedatos['ARREST_BORO'].value_counts( )

K    1496897
M    1452206
B    1232790
Q    1021719
S     192824
Name: ARREST_BORO, dtype: int64

In [32]:
# DE ACUERDO A https://www1.nyc.gov/site/nypd/bureaus/patrol/precincts-landing.page SOLO  son 77 precintos y en la base de datos hay 78
Basedatos['ARREST_PRECINCT'].value_counts( )

14     187460
75     182866
44     172301
40     158315
73     145516
        ...  
111     24230
94      24164
123     21410
22       6641
27          1
Name: ARREST_PRECINCT, Length: 78, dtype: int64

In [33]:
#Se localiza el precinto 27 ya que solo tiene 1 registro
Basedatos[Basedatos['ARREST_PRECINCT'] == 27]

ARREST_KEY  PD_CD                             PD_DESC  KY_CD  \
562258    32311380  511.0  CONTROLLED SUBSTANCE, POSSESSION 7  235.0   

              OFNS_DESC    LAW_CODE LAW_CAT_CD ARREST_BORO  ARREST_PRECINCT  \
562258  DANGEROUS DRUGS  PL 2200300          M           Q               27   

        JURISDICTION_CODE  ... PERP_RACE X_COORD_CD Y_COORD_CD  Latitude  \
562258                1.0  ...     BLACK        NaN        NaN       NaN   

        Longitude  Lon_Lat  ARREST_YEAR ARREST_DAY ARREST_MONTH      FECHA  
562258        NaN      NaN         2007         18            6 2007-06-18  

[1 rows x 22 columns]

In [34]:
#Se decide eliminar el precinto 27 debido a que se concideró como un dato atipico (pudo ser un error en la escritura del número), además cuenta con varios datos nulos
Basedatos.drop([562258])
Basedatos.head()

ARREST_KEY  PD_CD                                 PD_DESC  KY_CD  \
0     9930151  848.0  NY STATE LAWS,UNCLASSIFIED MISDEMEANOR  364.0   
1     9929960  905.0             INTOXICATED DRIVING,ALCOHOL  347.0   
2     9930303  793.0                    WEAPONS POSSESSION 3  118.0   
5     9929971  905.0             INTOXICATED DRIVING,ALCOHOL  347.0   
7     9930028  203.0                    TRESPASS 3, CRIMINAL  352.0   

                          OFNS_DESC    LAW_CODE LAW_CAT_CD ARREST_BORO  \
0  OTHER STATE LAWS (NON PENAL LAW)  LOC00000UM          M           M   
1    INTOXICATED & IMPAIRED DRIVING  VTL11920M3          M           K   
2                 DANGEROUS WEAPONS  PL 2650201          F           Q   
5    INTOXICATED & IMPAIRED DRIVING  VTL11920M2          M           K   
7                 CRIMINAL TRESPASS  PL 140100E          M           K   

   ARREST_PRECINCT  JURISDICTION_CODE  ...       PERP_RACE X_COORD_CD  \
0               18                0.0  ...           BLACK   988950.0   
1               84                0.0  ...           BLACK   988156.0   
2              101                0.0  ...           BLACK  1046242.0   
5               75                0.0  ...  WHITE HISPANIC  1017455.0   
7               73                2.0  ...  WHITE HISPANIC  1009450.0   

  Y_COORD_CD   Latitude  Longitude  \
0   216678.0  40.761414 -73.983035   
1   189697.0  40.687359 -73.985917   
2   157040.0  40.597506 -73.776773   
5   189330.0  40.686290 -73.880274   
7   182401.0  40.667298 -73.909163   

                                         Lon_Lat  ARREST_YEAR ARREST_DAY  \
0   POINT (-73.98303549199994 40.76141449100004)         2006          1   
1   POINT (-73.98591731099998 40.68735856600006)         2006          1   
2   POINT (-73.77677316599994 40.59750616900004)         2006          1   
5  POINT (-73.88027376799994 40.686289816000055)         2006          1   
7  POINT (-73.90916337699997 40.667297692000034)         2006          1   

  ARREST_MONTH      FECHA  
0            1 2006-01-01  
1            1 2006-01-01  
2            1 2006-01-01  
5            1 2006-01-01  
7            1 2006-01-01  

[5 rows x 22 columns]

In [35]:
Basedatos['PERP_SEX'].value_counts( )

M    4489620
F     906824
Name: PERP_SEX, dtype: int64

In [36]:
Basedatos['PERP_RACE'].value_counts( )

BLACK                             2623186
WHITE HISPANIC                    1398264
WHITE                              646961
BLACK HISPANIC                     436640
ASIAN / PACIFIC ISLANDER           226075
UNKNOWN                             51860
AMERICAN INDIAN/ALASKAN NATIVE      12101
OTHER                                1357
Name: PERP_RACE, dtype: int64

<p><a name="s3"></a></p>

# **TRATAMIENTO DE DATOS NULOS**


In [37]:
#Para ver los nulos
Basedatos.isnull().sum()

ARREST_KEY               0
PD_CD                  329
PD_DESC               9162
KY_CD                 9250
OFNS_DESC             9162
LAW_CODE               195
LAW_CAT_CD           21152
ARREST_BORO              8
ARREST_PRECINCT          0
JURISDICTION_CODE       10
AGE_GROUP               17
PERP_SEX                 0
PERP_RACE                0
X_COORD_CD               1
Y_COORD_CD               1
Latitude                 1
Longitude                1
Lon_Lat                  1
ARREST_YEAR              0
ARREST_DAY               0
ARREST_MONTH             0
FECHA                    0
dtype: int64

In [38]:
#Se calcula el porcentaje de nulos respecto a la base de datos
round(Basedatos.isnull().sum().sum()/Basedatos.size,22)*100

0.041517238860526405

In [39]:
#La variable KY_CD al tratarse de la categoría más general que la descripción PD se decide convertir los datos nulos en 000,
#de modo que deje de ser un dato nulo, al igual que la variable PD_CD.
#JURISDICTION_CODE se coloca 10000 ya que existe el codigo 0, para evitar confusiones se reemplaza por un número muy grande.

Basedatos.KY_CD.fillna(000,inplace=True)
Basedatos.PD_CD.fillna(000,inplace=True)
Basedatos.JURISDICTION_CODE.fillna(10000,inplace=True)

In [40]:
Basedatos.isnull().sum()

ARREST_KEY               0
PD_CD                    0
PD_DESC               9162
KY_CD                    0
OFNS_DESC             9162
LAW_CODE               195
LAW_CAT_CD           21152
ARREST_BORO              8
ARREST_PRECINCT          0
JURISDICTION_CODE        0
AGE_GROUP               17
PERP_SEX                 0
PERP_RACE                0
X_COORD_CD               1
Y_COORD_CD               1
Latitude                 1
Longitude                1
Lon_Lat                  1
ARREST_YEAR              0
ARREST_DAY               0
ARREST_MONTH             0
FECHA                    0
dtype: int64

In [41]:
#Para PD_DESC al tratarse de una descripción más detallada, se decide convertir los datos nulos en Desconocido,
#Al igual que las variables: OFNS_DESC, LAW_CODE, LAW_CAT_CD,ARREST_BORO y AGE_GROUP.

Basedatos.PD_DESC.fillna('Desconocido',inplace=True)
Basedatos.OFNS_DESC.fillna('Desconocido',inplace=True)
Basedatos.LAW_CODE.fillna('Desconocido',inplace=True)
Basedatos.LAW_CAT_CD.fillna('Desconocido',inplace=True)
Basedatos.ARREST_BORO.fillna('Desconocido',inplace=True)
Basedatos.AGE_GROUP.fillna('Desconocido',inplace=True)

In [42]:
# se observa la los datos nulos que hay en la variable X_COORD_CD
Basedatos[Basedatos['X_COORD_CD'].isnull()] 


ARREST_KEY  PD_CD                             PD_DESC  KY_CD  \
562258    32311380  511.0  CONTROLLED SUBSTANCE, POSSESSION 7  235.0   

              OFNS_DESC    LAW_CODE LAW_CAT_CD ARREST_BORO  ARREST_PRECINCT  \
562258  DANGEROUS DRUGS  PL 2200300          M           Q               27   

        JURISDICTION_CODE  ... PERP_RACE X_COORD_CD Y_COORD_CD  Latitude  \
562258                1.0  ...     BLACK        NaN        NaN       NaN   

        Longitude  Lon_Lat  ARREST_YEAR ARREST_DAY ARREST_MONTH      FECHA  
562258        NaN      NaN         2007         18            6 2007-06-18  

[1 rows x 22 columns]

In [43]:
## se observa los datos nulos que hay en la variable Latitude
Basedatos[Basedatos['Latitude'].isnull()]

ARREST_KEY  PD_CD                             PD_DESC  KY_CD  \
562258    32311380  511.0  CONTROLLED SUBSTANCE, POSSESSION 7  235.0   

              OFNS_DESC    LAW_CODE LAW_CAT_CD ARREST_BORO  ARREST_PRECINCT  \
562258  DANGEROUS DRUGS  PL 2200300          M           Q               27   

        JURISDICTION_CODE  ... PERP_RACE X_COORD_CD Y_COORD_CD  Latitude  \
562258                1.0  ...     BLACK        NaN        NaN       NaN   

        Longitude  Lon_Lat  ARREST_YEAR ARREST_DAY ARREST_MONTH      FECHA  
562258        NaN      NaN         2007         18            6 2007-06-18  

[1 rows x 22 columns]

In [44]:
# se confirma que los datos nulos de latitud y longitud ocurren en la misma fila, por lo tanto se decide eliminarlos.
Basedatos = Basedatos.drop([562258])

In [45]:
#No hay más registros de datos nulos
Basedatos.isnull().sum()

ARREST_KEY           0
PD_CD                0
PD_DESC              0
KY_CD                0
OFNS_DESC            0
LAW_CODE             0
LAW_CAT_CD           0
ARREST_BORO          0
ARREST_PRECINCT      0
JURISDICTION_CODE    0
AGE_GROUP            0
PERP_SEX             0
PERP_RACE            0
X_COORD_CD           0
Y_COORD_CD           0
Latitude             0
Longitude            0
Lon_Lat              0
ARREST_YEAR          0
ARREST_DAY           0
ARREST_MONTH         0
FECHA                0
dtype: int64

In [46]:
# Pasar las columnas a los formatos correctos
Basedatos['PD_CD'] = Basedatos['PD_CD'].apply(np.int64)
Basedatos['KY_CD'] = Basedatos['KY_CD'].apply(np.int64)

<p><a name="s3"></a></p>

# **Pregunta 1**


**¿Cuál ha sido la evolución diaria de la cantidad de arrestos?**
- a. a nivel general
- b. a nivel de ofensa (columna del nivel de ofensa: LAW_CAT_CD)
- c. a nivel de barrio (columna del barrio ARREST_BORO)

###*a) a nivel general*

In [47]:
#Se crea un DataFrame con todas las fechas para hacer la serie de tiempo
Seriefecha= pd.DataFrame(pd.date_range(start=Basedatos['FECHA'].min(), end=Basedatos['FECHA'].max())).rename(columns ={0:'FECHA'})
Seriefecha

FECHA
0    2006-01-01
1    2006-01-02
2    2006-01-03
3    2006-01-04
4    2006-01-05
...         ...
6020 2022-06-26
6021 2022-06-27
6022 2022-06-28
6023 2022-06-29
6024 2022-06-30

[6025 rows x 1 columns]

In [48]:
#Se unió la base de crimenes con las fechas
Seriefechacrimenes= pd.merge(Seriefecha, Basedatos, on ='FECHA', how ='left')

In [49]:
#Se hizo un groupby para contar los arrestos por día
Serie=Seriefechacrimenes.groupby(['FECHA'])[['ARREST_KEY']].count().reset_index()
Serie.head()

FECHA  ARREST_KEY
0 2006-01-01         537
1 2006-01-02         615
2 2006-01-03         898
3 2006-01-04        1227
4 2006-01-05        1379

In [50]:
#Antes de graficar se suaviza la serie
Serie2=Serie.copy()
Serie2.iloc[:,1:] = Serie2.rolling(window=30).mean()

In [51]:
fig = px.line(Serie2, x='FECHA', y ='ARREST_KEY', title = '<b>Evolución arrestos a nivel general<b>',
              color_discrete_sequence=px.colors.qualitative.Safe)

fig.update_layout(
    xaxis_title = 'Año',
    yaxis_title = 'Arrestos',
    template = 'simple_white',
    legend_title = '<b>Nivel de ofensa<b>',
    title_x = 0.5)
fig.show()

**Interpretación:** La cantidad general de arrestos tuvo el pico más grande entre el 2006 y el 2014, después de este año los arrestos empezaron a disminuir de manera constante antes de iniciar el año 2020 y finalmente se observa que desde el 2020, el número de arrestos ha ido incrementandose nuevamente.

###*b) a nivel de ofensa (columna del nivel de ofensa: LAW_CAT_CD)*

In [52]:
Basedatos['LAW_CAT_CD'].value_counts()

#Nivel de ofensa:
#felony: delito grave
#misdemeanor: delito menor
#violation: violación
#i: informático

M              3540015
F              1515259
V               293407
I                26610
Desconocido      21152
Name: LAW_CAT_CD, dtype: int64

In [53]:
#Se hizo un groupby para contar los arrestos a nivel de ofensa
ofensa=Seriefechacrimenes.groupby(['FECHA','LAW_CAT_CD'])[['ARREST_KEY']].count().reset_index()
ofensa.head()

FECHA   LAW_CAT_CD  ARREST_KEY
0 2006-01-01  Desconocido           1
1 2006-01-01            F         160
2 2006-01-01            I           6
3 2006-01-01            M         338
4 2006-01-01            V          32

In [54]:
#Se filtra por la ofensa Misdemeanor/delito menor
Misdemeanor=ofensa[ofensa['LAW_CAT_CD']=='M']

#Se crea una serie de tiempo nueva con el delito Misdemeanor 
ofensa1= pd.merge(Seriefecha,Misdemeanor, on ='FECHA', how ='left').rename(columns={'ARREST_KEY':'Misdemeanor'})
ofensa1.head()

FECHA LAW_CAT_CD  Misdemeanor
0 2006-01-01          M        338.0
1 2006-01-02          M        416.0
2 2006-01-03          M        555.0
3 2006-01-04          M        828.0
4 2006-01-05          M        863.0

In [55]:
#Se filtra por violación
Violacion=ofensa[ofensa['LAW_CAT_CD']=='V']
Violacion.head()

#Se crea una serie de tiempo para violación
ofensa2= pd.merge(Seriefecha,Violacion, on ='FECHA', how ='left').rename(columns={'ARREST_KEY':'Violacion'})
ofensa2.head()

FECHA LAW_CAT_CD  Violacion
0 2006-01-01          V       32.0
1 2006-01-02          V       29.0
2 2006-01-03          V       36.0
3 2006-01-04          V       50.0
4 2006-01-05          V       68.0

In [56]:
#Se filtra por delito grave
felony=ofensa[ofensa['LAW_CAT_CD']=='F']
felony.head()

#Serie de tiempo para delito grave
ofensa3= pd.merge(Seriefecha,felony, on ='FECHA', how ='left').rename(columns={'ARREST_KEY':'felony'})
ofensa3.head()

FECHA LAW_CAT_CD  felony
0 2006-01-01          F   160.0
1 2006-01-02          F   167.0
2 2006-01-03          F   300.0
3 2006-01-04          F   341.0
4 2006-01-05          F   441.0

In [57]:
#se filtra por delito informatico
informatica=ofensa[ofensa['LAW_CAT_CD']=='I']
informatica.head()

#serie de tiempo para delito informatico
ofensa4= pd.merge(Seriefecha,informatica, on ='FECHA', how ='left').rename(columns={'ARREST_KEY':'informaticos'})
ofensa4.head()

FECHA LAW_CAT_CD  informaticos
0 2006-01-01          I           6.0
1 2006-01-02          I           2.0
2 2006-01-03          I           3.0
3 2006-01-04          I           4.0
4 2006-01-05          I           4.0

In [58]:
#Se crea un dataframe nuevo con 5 columnas correspondientes a la cantidad de arrestos por cada tipo de delito
puntob = Seriefecha.copy()
puntob['Delito menor'] = ofensa1['Misdemeanor'].fillna(0)
puntob['Violacion'] = ofensa2['Violacion'].fillna(0)
puntob['Delito mayor'] = ofensa3['felony'].fillna(0)
puntob['Informatica'] = ofensa4['informaticos'].fillna(0)
puntob.head()

FECHA  Delito menor  Violacion  Delito mayor  Informatica
0 2006-01-01         338.0       32.0         160.0          6.0
1 2006-01-02         416.0       29.0         167.0          2.0
2 2006-01-03         555.0       36.0         300.0          3.0
3 2006-01-04         828.0       50.0         341.0          4.0
4 2006-01-05         863.0       68.0         441.0          4.0

In [59]:
#Antes de graficar se suaviza la serie
puntob2=puntob.copy()
puntob2.iloc[:,1:] = puntob2.rolling(window=30).mean()

In [60]:
fig = px.line(puntob2, x='FECHA', y =['Delito menor','Violacion','Delito mayor','Informatica'], title = '<b>Evolución arrestos por nivel de ofensa<b>',
              color_discrete_sequence=px.colors.qualitative.Bold)

fig.update_layout(
    xaxis_title = 'Año',
    yaxis_title = 'Arrestos',
    template = 'simple_white',
    legend_title = '<b>Nivel de ofensa<b>',
    title_x = 0.5)
fig.show()

**Interpretación:** El mayor nivel de ofensa que se presentan en los arrestos es del de delito menor (Misdemeanor) y este ha ido disminuyendo a partir del año 2015 aproximadamente hasta alcanzar a el delito mayor (Felony) en el año 2020, el cual se ha mantenido constante durante el tiempo. El tipo de ofensa violación, junto con informática no han crecido tan significativamente como los demás y son muy pocos los arrestos por estos dos.

###*c. a nivel de barrio (columna del barrio ARREST_BORO)*

In [61]:
Basedatos['ARREST_BORO'].value_counts()

#Barrio:
#K: Brooklyn
#M: Manhattan
#B: Bronx
#Q: Queens
#S: Staten Island

K              1496897
M              1452206
B              1232790
Q              1021718
S               192824
Desconocido          8
Name: ARREST_BORO, dtype: int64

In [62]:
#Se hace un groupby, agrupando por fecha y los barrios y contando los arrestos
barrio=Seriefechacrimenes.groupby(['FECHA','ARREST_BORO'])[['ARREST_KEY']].count().reset_index()
barrio.head()

FECHA ARREST_BORO  ARREST_KEY
0 2006-01-01           B         106
1 2006-01-01           K         148
2 2006-01-01           M         147
3 2006-01-01           Q         123
4 2006-01-01           S          13

In [63]:
#Se filtra por el barrio Brooklyn
Brooklyn=barrio[barrio['ARREST_BORO']=='K']

#Se crea una serie de tiempo nueva 
Brooklyn2= pd.merge(Seriefecha,Brooklyn, on ='FECHA', how ='left').rename(columns={'ARREST_KEY':'Brooklyn'})
Brooklyn2.head()

FECHA ARREST_BORO  Brooklyn
0 2006-01-01           K     148.0
1 2006-01-02           K     221.0
2 2006-01-03           K     236.0
3 2006-01-04           K     341.0
4 2006-01-05           K     354.0

In [64]:
#Se filtra por Manhattan
Manhattan=barrio[barrio['ARREST_BORO']=='M']

#Se crea una serie de tiempo
Manhattan2= pd.merge(Seriefecha,Manhattan, on ='FECHA', how ='left').rename(columns={'ARREST_KEY':'Manhattan'})
Manhattan2.head()

FECHA ARREST_BORO  Manhattan
0 2006-01-01           M      147.0
1 2006-01-02           M      131.0
2 2006-01-03           M      237.0
3 2006-01-04           M      302.0
4 2006-01-05           M      367.0

In [65]:
#Se filtra por Bronx
Bronx=barrio[barrio['ARREST_BORO']=='B']

#Se crea una serie de tiempo
Bronx2= pd.merge(Seriefecha,Bronx, on ='FECHA', how ='left').rename(columns={'ARREST_KEY':'Bronx'})
Bronx2.head()

FECHA ARREST_BORO  Bronx
0 2006-01-01           B  106.0
1 2006-01-02           B  140.0
2 2006-01-03           B  204.0
3 2006-01-04           B  286.0
4 2006-01-05           B  379.0

In [66]:
#Se filtra por Queens
Queens=barrio[barrio['ARREST_BORO']=='Q']

#Se crea una serie de tiempo
Queens2= pd.merge(Seriefecha,Queens, on ='FECHA', how ='left').rename(columns={'ARREST_KEY':'Queens'})
Queens2.head()

FECHA ARREST_BORO  Queens
0 2006-01-01           Q   123.0
1 2006-01-02           Q   101.0
2 2006-01-03           Q   183.0
3 2006-01-04           Q   251.0
4 2006-01-05           Q   226.0

In [67]:
#Se filtra por Staten Island
StatenIsland=barrio[barrio['ARREST_BORO']=='S']

#Se crea una serie de tiempo
StatenIsland2= pd.merge(Seriefecha,StatenIsland, on ='FECHA', how ='left').rename(columns={'ARREST_KEY':'Staten Island'})
StatenIsland2.head()

FECHA ARREST_BORO  Staten Island
0 2006-01-01           S           13.0
1 2006-01-02           S           22.0
2 2006-01-03           S           38.0
3 2006-01-04           S           47.0
4 2006-01-05           S           53.0

In [68]:
#Se crea un dataframe nuevo con 5 columnas correspondientes a la cantidad de arrestos por cada tipo de delito
puntoc = Seriefecha.copy()
puntoc['Brooklyn'] = Brooklyn2['Brooklyn'].fillna(0)
puntoc['Manhattan'] = Manhattan2['Manhattan'].fillna(0)
puntoc['Bronx'] = Bronx2['Bronx'].fillna(0)
puntoc['Queens'] = Queens2['Queens'].fillna(0)
puntoc['Staten Island'] = StatenIsland2['Staten Island'].fillna(0)
puntoc.head()

FECHA  Brooklyn  Manhattan  Bronx  Queens  Staten Island
0 2006-01-01     148.0      147.0  106.0   123.0           13.0
1 2006-01-02     221.0      131.0  140.0   101.0           22.0
2 2006-01-03     236.0      237.0  204.0   183.0           38.0
3 2006-01-04     341.0      302.0  286.0   251.0           47.0
4 2006-01-05     354.0      367.0  379.0   226.0           53.0

In [69]:
puntoc2=puntoc.copy()

In [70]:
#Antes de graficar se suaviza la serie
puntoc2.iloc[:,1:] = puntoc2.rolling(window=30).mean()

In [71]:
#Se crea la gráfica
fig = px.line(puntoc2, x='FECHA', y =['Brooklyn','Manhattan','Bronx','Queens','Staten Island'], title = '<b>Evolución arrestos por barrio<b>')

fig.update_layout(
    xaxis_title = 'Año',
    yaxis_title = 'Arrestos',
    template = 'simple_white',
    legend_title = '<b>Nivel de ofensa<b>',
    title_x = 0.5)
fig.show()

**Interpretación:** 
- La mayor cantidad de arrestos a través del tiempo se han presentado en el barrio Brooklyn aunque este tiene cantidad de arrestos similares a el barrio Manhattan, por lo cual se podría afirmar que ambos barrios son los más peligrosos de la ciudad de Nueva York. Por otro lado, los barrios Bronx y Queens se comportan de manera similiar a través de los años, mientras que el barrio Staten Island se ha mantenido constante en el número de arrestos siendo estos los más bajos, es por esto que este barrio es el más seguro en términos de arrestos.
- Otro punto a destacar es que los barrios Brooklyn, Manhattan, Bronx y Queens durante el año 2020 disminuyeron los números de arrestos considerablemente.


<p><a name="s3"></a></p>

# **Pregunta 2**


**Ver la evolución general agrupando por semana y trimestre ¿Encuentra algo que le llame la
atención?**

In [72]:
Seriecrimen=Seriefechacrimenes.groupby(['FECHA'])[['ARREST_KEY']].count().reset_index()

In [73]:
#Serie de tiempo por semana
Crimensemanal=Seriecrimen.resample('W-Mon',on ='FECHA').sum().reset_index()
Crimensemanal.head()

FECHA  ARREST_KEY
0 2006-01-02        1152
1 2006-01-09        7170
2 2006-01-16        7117
3 2006-01-23        6090
4 2006-01-30        7458

In [74]:
#Se crea la gráfica
fig = px.line(Crimensemanal, x='FECHA', y ='ARREST_KEY', title = '<b>Evolución arrestos por semana<b>')

fig.update_layout(
    xaxis_title = 'Año',
    yaxis_title = 'Arrestos',
    template = 'simple_white',
    legend_title = '<b>Nivel de ofensa<b>',
    title_x = 0.5)
fig.show()

**Interpretación:** En los últimos meses de cada año (noviembre y diciembre) se puede ver a través del grafico como los arrestos en general van disminuyendo, mientras que en los meses marzo, abril y mayo, la cantidad de arrestos empieza a incrementarse considerablemente.

In [75]:
Seriecrimen=Seriefechacrimenes.groupby(['FECHA'])[['ARREST_KEY']].count().reset_index()

In [76]:
#Serie de tiempo de arrestos por trimestre
Crimentrimestre=Seriecrimen.resample('3M',on ='FECHA').sum().reset_index() # Reagrupa por mes y suma los valores que comprende este periodo

In [77]:
Crimentrimestre.head()

FECHA  ARREST_KEY
0 2006-01-31       29984
1 2006-04-30       93485
2 2006-07-31       93364
3 2006-10-31       95543
4 2007-01-31       95935

In [78]:
#Se crea la gráfica
fig = px.line(Crimentrimestre, x='FECHA', y ='ARREST_KEY', title = '<b>Evolución arrestos por trimestre<b>')

fig.update_layout(
    xaxis_title = 'Año',
    yaxis_title = 'Arrestos',
    template = 'simple_white',
    legend_title = '<b>Nivel de ofensa<b>',
    title_x = 0.5)
fig.show()

**Interpretación:** En la grafica trimestral se observa un comportamiento similar a la grafica anterior, en donde se confirma que aproximadamente entre los meses de abril, la cantidad de arrestos alcanza su punto máximo a través de los años, mientras que en meses como enero esta cantidad tiende a disminuir.

<p><a name="s3"></a></p>

# **Pregunta 3**



**¿Cuál es el barrio en el que se presentan más y menos arrestos por población y área?**

In [79]:
Seriefechacrimenes['ARREST_BORO'].value_counts()
#B(Bronx), S(Staten Island), K(Brooklyn), M(Manhattan), Q(Queens)

K              1496897
M              1452206
B              1232790
Q              1021718
S               192824
Desconocido          8
Name: ARREST_BORO, dtype: int64

In [80]:
Seriebarrio=Seriefechacrimenes.groupby(['ARREST_BORO'])[['FECHA']].count().reset_index()
Seriebarrio

ARREST_BORO    FECHA
0            B  1232790
1  Desconocido        8
2            K  1496897
3            M  1452206
4            Q  1021718
5            S   192824

In [81]:
Seriebarrio['Indice_arrestos_personas']=0

In [82]:
#Datos sacados de Wikipedia al año 2020 (población de cada barrio)
Seriebarrio.loc[0,'Indice_arrestos_personas']=Seriebarrio.loc[0,'FECHA']/1427000 #Bronx
Seriebarrio.loc[2,'Indice_arrestos_personas']=Seriebarrio.loc[2,'FECHA']/2577000 #Brooklyn
Seriebarrio.loc[3,'Indice_arrestos_personas']=Seriebarrio.loc[3,'FECHA']/1629000 #Manhattan
Seriebarrio.loc[4,'Indice_arrestos_personas']=Seriebarrio.loc[4,'FECHA']/2271000 #Queens
Seriebarrio.loc[5,'Indice_arrestos_personas']=Seriebarrio.loc[5,'FECHA']/475596 #Staten Island

#NOTA: no se toma en cuenta el barrio desconocido porque no se puede buscar información sobre la población

In [83]:
#Se crea un diccionario para reemplazar las iniciales por el nombre del barrio
dicbarrio={'B':'Bronx','S':'Staten Island','K':'Brooklyn','Q':'Queens','M':'Manhattan'}

Seriebarrio['ARREST_BORO']=Seriebarrio['ARREST_BORO'].replace(dicbarrio)
Seriebarrio

ARREST_BORO    FECHA  Indice_arrestos_personas
0          Bronx  1232790                  0.863903
1    Desconocido        8                  0.000000
2       Brooklyn  1496897                  0.580868
3      Manhattan  1452206                  0.891471
4         Queens  1021718                  0.449898
5  Staten Island   192824                  0.405437

In [84]:
Seriebarrio = Seriebarrio[Seriebarrio['ARREST_BORO']!='Desconocido']

In [85]:
#Se organiza el dataframe de mayor indice a menor
Seriebarrio=Seriebarrio.reset_index().sort_values('Indice_arrestos_personas',ascending=False)

In [86]:
#Se crea la gráfica
fig=px.bar(Seriebarrio,x='ARREST_BORO', y='Indice_arrestos_personas', title='<b>Indice de arrestos históricos por cantidad de personas<b>')
fig.update_layout(
    xaxis_title='Barrio',
    yaxis_title = '% de arrestos',
    template='simple_white',
    title_x=0.5)

In [87]:
Seriebarrio['Indice_arrestos_area']=0

In [88]:
Seriebarrio

index    ARREST_BORO    FECHA  Indice_arrestos_personas  \
2      3      Manhattan  1452206                  0.891471   
0      0          Bronx  1232790                  0.863903   
1      2       Brooklyn  1496897                  0.580868   
3      4         Queens  1021718                  0.449898   
4      5  Staten Island   192824                  0.405437   

   Indice_arrestos_area  
2                     0  
0                     0  
1                     0  
3                     0  
4                     0

In [89]:
#Informacion tomada de Wikipedia en Km^2 (área de cada barrio)
Seriebarrio.loc[0,'Indice_arrestos_area']=Seriebarrio.loc[0,'FECHA']/110 #Bronx
Seriebarrio.loc[1,'Indice_arrestos_area']=Seriebarrio.loc[1,'FECHA']/180 #Brooklyn
Seriebarrio.loc[2,'Indice_arrestos_area']=Seriebarrio.loc[2,'FECHA']/59 #Manhattan
Seriebarrio.loc[3,'Indice_arrestos_area']=Seriebarrio.loc[3,'FECHA']/280 #Queens
Seriebarrio.loc[4,'Indice_arrestos_area']=Seriebarrio.loc[4,'FECHA']/152 #Staten Island

#NOTA: no se toma en cuenta el barrio desconocido porque no se puede buscar información sobre el área

In [90]:
Seriebarrioarea=Seriebarrio[['ARREST_BORO','Indice_arrestos_area']]
Seriebarrioarea

ARREST_BORO  Indice_arrestos_area
2      Manhattan          24613.661017
0          Bronx          11207.181818
1       Brooklyn           8316.094444
3         Queens           3648.992857
4  Staten Island           1268.578947

In [91]:
#Se organiza el dataframe de mayor indice a menor
Seriebarrioarea=Seriebarrioarea.reset_index().sort_values('Indice_arrestos_area',ascending=False)

In [92]:
#Se crea la gráfica
fig=px.bar(Seriebarrioarea,x='ARREST_BORO', y='Indice_arrestos_area', title='<b>Indice de arrestos históricos por área<b>')
fig.update_layout(
    xaxis_title='Barrio',
    yaxis_title = '# de arrestos',
    template='simple_white',
    title_x=0.5)

**Interpretación:** Se observa en la gráfica el indice de arrestos por personas que el barrio Manhattan es el más inseguro, donde se presentan más delitos, seguido por el Bronx y Brooklyn, en cambio el más seguro es Staten Island. Al igual que en la gráfica de indice de arrestos por área, en donde el barrio Queens y Staten Island a diferencia de los demás la porporción de arrestos no es tan alta.

<p><a name="s3"></a></p>

# **Pregunta 4**


**¿Cómo se ve el comportamiento de estos barrios por día de la semana y día del mes? ¿Se
presentan más arrestos en las quincenas o los fines de semana?**

In [93]:
Seriebarrioperiodo=Seriebarrio.copy()

In [94]:
#Se hace agrupa por fecha y se cuenta el promedio de los arrestos
Seriebarrioperiodo=Seriecrimen.groupby(['FECHA'])[['ARREST_KEY']].mean().reset_index()
Seriebarrioperiodo.head()

FECHA  ARREST_KEY
0 2006-01-01       537.0
1 2006-01-02       615.0
2 2006-01-03       898.0
3 2006-01-04      1227.0
4 2006-01-05      1379.0

In [95]:
#Cambiamos cada fecha al dia de la semana
Seriebarrioperiodo['DIA']=Seriebarrioperiodo['FECHA'].dt.weekday
Seriebarrioperiodo['MES']=Seriebarrioperiodo['FECHA'].dt.month

In [96]:
#Lo agrupamos por días y se les sacó la media para realizar la estadística del promedio de arrestos por día
Seriebarrioperiodo=Seriebarrioperiodo.groupby(['DIA','MES'])[['ARREST_KEY']].count().reset_index().rename(columns={'ARREST_KEY':'ARRESTOS'})

In [97]:
#Se crea un diccionario para reemplazar los numeros de los días por:
#0=lunes, 1=martes, 2=miercoles, 3=jueves, 4=viernes, 5=sabado, 6=domingo

dias={0:'Lunes',1:'Martes',2:'Miercoles',3:'Jueves',4:'Viernes',5:'Sabado',6:'Domingo'}

#Se crea un diccionario para reemplazar los numeros de los meses por:
#1=Enero, 2=Febrero, 3=Marzo, 4=Abril, 5=Mayo, 6=Junio, 7=Julio, 8=Agosto, 9=Septiembre, 10=Octubre, 11=Noviembre, 12=Diciembre

meses ={1:'Enero', 2:'Febrero', 3:'Marzo', 4:'Abril', 5:'Mayo', 6:'Junio', 7:'Julio', 8:'Agosto', 9:'Septiembre', 10:'Octubre', 11:'Noviembre', 12:'Diciembre'}

Seriebarrioperiodo['DIA']=Seriebarrioperiodo['DIA'].replace(dias)
Seriebarrioperiodo['MES']=Seriebarrioperiodo['MES'].replace(meses)

In [98]:
fig = px.line(Seriebarrioperiodo, x="DIA", y="ARRESTOS", facet_col="MES", facet_col_wrap=4, title='<b>Evolución de arrestos por día de la semana<b>',
              category_orders={"DIA": ["Lunes", "Martes", "Miercoles", "Jueves","Viernes","Sabado","Domingo"]},
              color = 'MES',
              color_discrete_sequence=px.colors.qualitative.Set2)

fig.update_layout(template='simple_white',
                  title_x=0.5)
fig.show()

**Interpretación:** En la mayoría de meses, la mayor catidad de arrestos ocurren los días martes, miercoles y jueves, es decir, a mitad de semana. Sin embargo, en algunos meses estos días no son tan significativos porque ocurren la misma cantidad de arrestos como en junio y octubre. Los días en los que menos arrestos hay, en promedio son los domingos y lunes, es probable que esto se deba a que son días en los que las personas tienden a quedarse en casa.

In [99]:
Seriebarrioperiodomes=Seriecrimen.copy()

In [100]:
#convertimos la fecha a día del mes
Seriebarrioperiodomes['DIA']=Seriebarrioperiodomes['FECHA'].dt.day
Seriebarrioperiodomes['MES']=Seriebarrioperiodomes['FECHA'].dt.month
Seriebarrioperiodomes['AÑO']=Seriebarrioperiodomes['FECHA'].dt.year

In [101]:
#agrupamos y sacamos la media
Seriebarrioperiodomes=Seriebarrioperiodomes.groupby(['DIA','MES'])[['ARREST_KEY']].mean().reset_index().rename(columns={'ARREST_KEY':'ARRESTOS'})

In [102]:
#Se crea un diccionario para reemplazar los numeros de los meses por:
#1=Enero, 2=Febrero, 3=Marzo, 4=Abril, 5=Mayo, 6=Junio, 7=Julio, 8=Agosto, 9=Septiembre, 10=Octubre, 11=Noviembre, 12=Diciembre

meses ={1:'Enero', 2:'Febrero', 3:'Marzo', 4:'Abril', 5:'Mayo', 6:'Junio', 7:'Julio', 8:'Agosto', 9:'Septiembre', 10:'Octubre', 11:'Noviembre', 12:'Diciembre'}

Seriebarrioperiodomes['MES']=Seriebarrioperiodomes['MES'].replace(meses)

In [103]:
fig = px.line(Seriebarrioperiodomes, x="DIA", y="ARRESTOS", facet_col="MES", facet_col_wrap=4, title='<b>Evolución de arrestos por día del mes<b>',
              color = 'MES',
              color_discrete_sequence=px.colors.qualitative.Set2)

fig.update_layout(template='simple_white',
                  title_x=0.5)
fig.show()


**Interpretación:** Cuando se está acabando el mes, aproximadamente a partir de los días 20, la cantidad de arrestos disminuye. Además, en algunos meses como febrero, marzo, abril, agosto, octubre y diciembre en los días 15 también tiende a disminuir los arrestos.

<p><a name="s3"></a></p>

# **Pregunta 5**


**¿Cuál es el pareto de arrestos por ofensa general (OFNS_DESC)?**

In [104]:
#Se crea el dataframe para hacer el pareto, agrupando por la ofensa
df0 = Basedatos.groupby(['OFNS_DESC'])[['ARREST_KEY']].count().sort_values('ARREST_KEY', ascending = False).rename(columns={'ARREST_KEY':'counts'}).head(10)
df0['ratio'] = df0.apply(lambda x: x.cumsum()/df0['counts'].sum())
df0

counts     ratio
OFNS_DESC                                         
DANGEROUS DRUGS                  1102942  0.314081
ASSAULT 3 & RELATED OFFENSES      557448  0.472824
OTHER OFFENSES RELATED TO THEFT   303909  0.559368
PETIT LARCENY                     244410  0.628967
OTHER STATE LAWS                  241600  0.697767
FELONY ASSAULT                    234767  0.764621
VEHICLE AND TRAFFIC LAWS          213023  0.825283
DANGEROUS WEAPONS                 208472  0.884649
CRIMINAL TRESPASS                 202830  0.942408
MISCELLANEOUS PENAL LAW           202242  1.000000

In [105]:
# importar paquete
import plotly.graph_objects as go 


# definir figura
fig = go.Figure([go.Bar(x=df0.index, y=df0['counts'], yaxis='y1', name='ARREST_KEY'),
                 go.Scatter(x=df0.index, y=df0['ratio'], yaxis='y2', name='ARREST_KEY', hovertemplate='%{y:.1%}', marker={'color': '#000000'})])

# agregar detalles
fig.update_layout(template='plotly_white', showlegend=False, hovermode='x', bargap=.3,
                  title={'text': '<b>Top 10 Arrestos por ofensa general<b>', 'x': .5}, 
                  yaxis={'title': 'Arrestos'},
                  yaxis2={'rangemode': "tozero", 'overlaying': 'y', 'position': 1, 'side': 'right', 'title': 'ratio', 'tickvals': np.arange(0, 1.1, .2), 'tickmode': 'array', 'ticktext': [str(i) + '%' for i in range(0, 101, 20)]})

fig.show()

**Interpretación:** Los arrestos por ofensa general más frecuentes son por drogas peligrosas, asaltos 3 y delitos relacionados, y otros delitos relacionados con el robo. Teniendo un porcentaje de frecuencia en este top 10 del 55,9%. Estos delitos estan relacionados con el consumo de drogas y los robos en general en donde se utilizan instrumentos peligrosos como armas, con el fin de causar daño a las víctimas.

<p><a name="s3"></a></p>

# **Pregunta 6**


**En el tipo de ofensa que más se presenta en los arrestos (posición 1 del pareto anterior)
analizar la descripción más detallada del tipo de ofensa (PD_DESC)**

In [106]:
#Se crea un groupby con la ofensa 'DANGEROUS DRUGS' y sus categorias
drogas = Basedatos[Basedatos['OFNS_DESC'].isin(['DANGEROUS DRUGS'])]
drogas = drogas.groupby(['OFNS_DESC','PD_DESC'])[['ARREST_KEY']].count().reset_index().sort_values('ARREST_KEY', ascending = False).head(10)

In [107]:
# crear gráfica
fig = px.bar(drogas, x='PD_DESC', y='ARREST_KEY', barmode ='group', title ='<b>Información detallada de arrestos por drogas peligrosas<b>',
             color_discrete_sequence=px.colors.qualitative.Set3, text_auto='.2s')

# agregar detalles a la gráfica
fig.update_layout(
    xaxis_title = 'Descripción',
    yaxis_title = 'Arrestos',
    template = 'simple_white',
    title_x = 0.5)
fig.update_traces(textfont_size=12, textangle=0, textposition="outside", cliponaxis=False)
fig.show()

**Interpretacion:** La ofensa drogas peligrosas que se presenta con más frecuencia en los arrestos de la ciudad de Nueva York, está relacionada como se ilustra en este top 10, por el consumo, posesión, venta y control de estás drogas y sustancias psicoactivas, como lo son en su mayoría la marihuana con más de 420.000 arrestos. 

<p><a name="s3"></a></p>

# **Pregunta 7**


**Por cada barrio ¿Cuáles son los códigos de Ley penal que más se presentan (Top 5)? y ¿A qué
se refieren?**

In [108]:
Basedatos['ARREST_BORO'].value_counts()
# B(Bronx), S(Staten Island), K(Brooklyn), M(Manhattan), Q(Queens)

K              1496897
M              1452206
B              1232790
Q              1021718
S               192824
Desconocido          8
Name: ARREST_BORO, dtype: int64

In [109]:
Basedatos['LAW_CODE'].value_counts()

PL 1200001    423617
PL 2211001    405856
PL 1651503    298297
PL 2200300    296900
PL 1552500    244243
               ...  
VTL1192UFS         1
PL 2753303         1
PL 125271C         1
PL 2753302         1
PL 12005SH         1
Name: LAW_CODE, Length: 2492, dtype: int64

###**Bronx**

In [110]:
#Dataframe para el primer barrio B(Bronx)
bronx = Basedatos[Basedatos['ARREST_BORO'].isin(['K'])]
bronx = bronx.groupby(['LAW_CODE','OFNS_DESC'])[['ARREST_KEY']].count().reset_index().sort_values('ARREST_KEY', ascending = False).head(5)
bronx

LAW_CODE                        OFNS_DESC  ARREST_KEY
1260  PL 2211001                  DANGEROUS DRUGS      131664
272   PL 1200001     ASSAULT 3 & RELATED OFFENSES      120702
1155  PL 2200300                  DANGEROUS DRUGS       91773
200   LOC000000V                 OTHER STATE LAWS       81233
873   PL 1651503  OTHER OFFENSES RELATED TO THEFT       76398

**Interpretación:** Se observa en el top 5 en el Bronx las leyes que corresponden a los delitos de: dangerous drugs, assault 3 & related offenses, other state laws y other offenses related to theft. (Drogas peligrosas, asalto 3 & delitos relacionados, otras leyes estatales y otros delitos relacionados con el robo)


In [111]:
#grafica para el Bronx
fig = px.pie(bronx , values = 'ARREST_KEY', names = 'LAW_CODE', title = '<b>Códigos de Ley penal que más se presentan en el Bronx<b>',
             color_discrete_sequence=px.colors.qualitative.Set3)

# agregar detalles a la gráfica:
fig.update_layout(
    template = 'simple_white',
    legend_title = '<b>Códigos de ley penal<b>',
    title_x = 0.5)

fig.show()

###**Staten Island**

In [112]:
#Dataframe para el segundo barrio S(Staten Island)
staten = Basedatos[Basedatos['ARREST_BORO'].isin(['S'])]
staten = staten.groupby(['LAW_CODE','OFNS_DESC'])[['ARREST_KEY']].count().reset_index().sort_values('ARREST_KEY', ascending = False).head(5)
staten

LAW_CODE                     OFNS_DESC  ARREST_KEY
121  PL 1200001  ASSAULT 3 & RELATED OFFENSES       17256
706  PL 2200300               DANGEROUS DRUGS       14995
786  PL 2211001               DANGEROUS DRUGS       11382
431  PL 1552500                 PETIT LARCENY       10736
730  PL 2201601               DANGEROUS DRUGS        6464

**Interpretación:** Se observa en el top 5 en Staten Island las leyes que corresponden a los delitos de: Assault 3 & relates offenses, dangerous drugs y petit larceny. (Agresión 3 & delitos relacionados, drogas peligrosas y pequeño hurto). Siendo agresión 3 & delitos relacionados el de mayor proporción con un 28,4% de participación.

In [113]:
#grafica para Staten Island
fig = px.pie(staten , values = 'ARREST_KEY', names = 'LAW_CODE', title = '<b>Códigos de Ley penal que más se presentan en Staten Island<b>',
             color_discrete_sequence=px.colors.qualitative.Set2)

# agregar detalles a la gráfica:
fig.update_layout(
    template = 'simple_white',
    legend_title = '<b>Códigos de ley penal<b>',
    title_x = 0.5)

fig.show()

###**Brooklyn**

In [114]:
#Dataframe para el tercer barrio K(Brooklyn)
brooklyn = Basedatos[Basedatos['ARREST_BORO'].isin(['K'])]
brooklyn = brooklyn.groupby(['LAW_CODE','OFNS_DESC'])[['ARREST_KEY']].count().reset_index().sort_values('ARREST_KEY', ascending = False).head(5)
brooklyn

LAW_CODE                        OFNS_DESC  ARREST_KEY
1260  PL 2211001                  DANGEROUS DRUGS      131664
272   PL 1200001     ASSAULT 3 & RELATED OFFENSES      120702
1155  PL 2200300                  DANGEROUS DRUGS       91773
200   LOC000000V                 OTHER STATE LAWS       81233
873   PL 1651503  OTHER OFFENSES RELATED TO THEFT       76398

**Interpretación:** Se observa en el top 5 en Brooklyn las leyes que corresponden a los delitos de: Dangerous drugs, assault 3 & related offenses, other state laws y other offenses related to theft. (Drogas peligrosas, asalto 3 & delitos relacionados, otras leyes estatales y otros delitos relacionados con el robo). Aunque las proporciones de estos delitos son similares, predominan en mayor cantidad el delito de drogas peligrosas y asalto 3 & delitos relacionados, con 26,2% y 24,1% respectivamente.

In [115]:
#grafica para Brooklyn
fig = px.pie(brooklyn , values = 'ARREST_KEY', names = 'LAW_CODE', title = '<b>Códigos de Ley penal que más se presentan en Brooklyn<b>',
             color_discrete_sequence=px.colors.qualitative.Pastel)

# agregar detalles a la gráfica:
fig.update_layout(
    template = 'simple_white',
    legend_title = '<b>Códigos de ley penal<b>',
    title_x = 0.5)

fig.show()

###**Manhattan**

In [116]:
#Dataframe para el cuarto barrio M(Manhattan)
manhattan = Basedatos[Basedatos['ARREST_BORO'].isin(['M'])]
manhattan = manhattan.groupby(['LAW_CODE','OFNS_DESC'])[['ARREST_KEY']].count().reset_index().sort_values('ARREST_KEY', ascending = False).head(5)
manhattan

LAW_CODE                        OFNS_DESC  ARREST_KEY
913   PL 1651503  OTHER OFFENSES RELATED TO THEFT      109885
1319  PL 2211001                  DANGEROUS DRUGS       90209
782   PL 1552500                    PETIT LARCENY       90081
317   PL 1200001     ASSAULT 3 & RELATED OFFENSES       82733
242   LOC000000V                 OTHER STATE LAWS       82283

**Interpretación:** Se observa en el top 5 de Manhattan las leyes que corresponden a los delitos de: Other offenses related to theft, dangerous drugs, petit larceny, assaul 3 & related offenses y other state laws. (Otros delitos relacionados con el robo, drogas peligrosas, hurto pequeño, asalto 3 &  delitos relacionados y otras leyes estatales). Estos 5 delitos se presentaron en proporciones similares, cada un con un 20% aproximadamente.

In [117]:
#grafica para Manhattan
fig = px.pie(manhattan , values = 'ARREST_KEY', names = 'LAW_CODE', title = '<b>Códigos de Ley penal que más se presentan en Manhattan<b>',
             color_discrete_sequence=px.colors.qualitative.Pastel1)

# agregar detalles a la gráfica:
fig.update_layout(
    template = 'simple_white',
    legend_title = '<b>Códigos de ley penal<b>',
    title_x = 0.5)

fig.show()

###**Queens**

In [118]:
##Dataframe para el cuarto barrio Q(Queens)
queens = Basedatos[Basedatos['ARREST_BORO'].isin(['Q'])]
queens = queens.groupby(['LAW_CODE','OFNS_DESC'])[['ARREST_KEY']].count().reset_index().sort_values('ARREST_KEY', ascending = False).head(5)
queens

LAW_CODE                     OFNS_DESC  ARREST_KEY
248   PL 1200001  ASSAULT 3 & RELATED OFFENSES       92390
1187  PL 2211001               DANGEROUS DRUGS       59631
706   PL 1552500                 PETIT LARCENY       43067
195   LOC000000V              OTHER STATE LAWS       42269
1767  VTL051101A      OTHER TRAFFIC INFRACTION       42010

**Interpretación:** Se observa en el top 5 de Queens las leyes que corresponden a los delitos de: Assalult 3 & related offenses, dangerous drugs, petit larceny, other state laws y other traffic infraction. (Agresión 3 & delitos relacionados, drogas peligrosas, hurto pequeño, otras leyes estatales y otras infracciones de tráfico). Con un 33,1% para el delito de gresión 3 & delitos relacionados.

In [119]:
#grafica para Queens
fig = px.pie(queens , values = 'ARREST_KEY', names = 'LAW_CODE', title = '<b>Códigos de Ley penal que más se presentan en Queens<b>',
             color_discrete_sequence=px.colors.qualitative.Safe)

# agregar detalles a la gráfica:
fig.update_layout(
    template = 'simple_white',
    legend_title = '<b>Códigos de ley penal<b>',
    title_x = 0.5)

fig.show()

**Interpretación general:** En los 5 barrios analizados anteriormente, se presentaron incumplimiento de las leyes que corresponden a delitos similares,  entre ellos están los más destacados que fueron agresión 3 & delitos relacionados (PL 1200001) y drogas peligrosas (PL 2211001).

<p><a name="s3"></a></p>

# **Pregunta 8**


**¿Cuál es el precinto más peligroso? ¿Este análisis cambia al agregar el mes del año?**

In [120]:
#Se agrupa por precinto
precinto= Basedatos.groupby(['ARREST_PRECINCT'])[['ARREST_KEY']].count().sort_values('ARREST_KEY', ascending = False).reset_index().head(10)
precinto

ARREST_PRECINCT  ARREST_KEY
0               14      187460
1               75      182866
2               44      172301
3               40      158315
4               73      145516
5               46      133638
6               43      131732
7               52      126858
8               25      120020
9              103      112109

In [121]:
precinto['ARREST_PRECINCT']=precinto['ARREST_PRECINCT'].astype(str)

In [122]:
# crear gráfica
fig = px.bar(precinto, x='ARREST_PRECINCT', y='ARREST_KEY', title ='<b>Precintos más peligrosos<b>')

# agregar detalles a la gráfica
fig.update_layout(
    xaxis_title = 'Precintos',
    yaxis_title = 'Cantidad arrestos',
    template = 'simple_white',
    title_x = 0.5)

fig.show()


**Interpretación:** Se evidencia que el precinto #14 es el que presenta más delitos con 187460 arrestos, seguido del precinto #75 con 182866 arrestos.

In [123]:
#Se agrupa por fecha y precinto
precinto=Seriefechacrimenes.groupby(['FECHA','ARREST_PRECINCT'])[['ARREST_KEY']].count().reset_index()
precinto.head()

FECHA  ARREST_PRECINCT  ARREST_KEY
0 2006-01-01              1.0           1
1 2006-01-01              5.0           5
2 2006-01-01              6.0           2
3 2006-01-01              7.0          12
4 2006-01-01              9.0          16

In [124]:
# Se saca el mes
precinto['FECHA']=precinto['FECHA'].dt.month

In [125]:
#Se crea un diccionario para reemplazar por el nombre del mes
meses1={1:'Enero',2:'Febrero',3:'Marzo',4:'Abril',5:'Mayo',6:'Junio',7:'Julio',8:'Agosto',9:'Septiembre',10:'Octubre',11:'Noviembre',12:'Diciembre'}

In [126]:
precinto['FECHA']=precinto['FECHA'].replace(meses1)
precinto.head()

FECHA  ARREST_PRECINCT  ARREST_KEY
0  Enero              1.0           1
1  Enero              5.0           5
2  Enero              6.0           2
3  Enero              7.0          12
4  Enero              9.0          16

In [127]:
# redondear los numeros de los precintos y despues pasarlos a str ya que son categorias
precinto['ARREST_PRECINCT'] = precinto['ARREST_PRECINCT'].apply(np.int64)
precinto['ARREST_PRECINCT'] = precinto['ARREST_PRECINCT'].astype(str)

In [128]:
base = precinto.groupby(['ARREST_PRECINCT','FECHA'])[['ARREST_KEY']].sum().reset_index().sort_values('ARREST_KEY',ascending=False).rename(columns={'ARREST_KEY':'Arrestos','FECHA':'Mes','ARREST_PRECINCT':'Precinto'})

In [129]:
# Crear grafica
fig = px.bar(base, x="Precinto", y="Arrestos", facet_col="Mes", facet_col_wrap=3, color='Precinto',
             title = '<b>Precintos en cada mes<b>',
             category_orders={'Mes':['Enero','Febrero','Marzo','Abril','Mayo','Junio','Julio','Agosto','Septiembre','Octubre','Noviembre','Diciembre']},
             height=1200)

# agregar detalles a la gráfica
fig.update_layout(
    template = 'simple_white',
    title_x = 0.5,
    legend_title = '<b>Precinto<b>')
fig.show()

**Interpretación:**se observa que en los mesos (Enero,Febrero,Marzo,Octubre,Noviembre,Diciembre), al final y principio del año, el precinto 14 es el que permanece en primer lugar en cantidad de arrestos, mientras que el precinto 75 permanece en primer lugar en los meses restantes (mitad de año).

<p><a name="s3"></a></p>

# **Pregunta 9**



**¿Cómo ha sido la evolución del precinto (ARREST_PRECINCT) más peligroso? ¿Ha mejorado o empeorado?**

In [130]:
#Precinto más peligroso 14 con 187400 arrestos en total
precinto_peligroso=Seriefechacrimenes.groupby(['FECHA','ARREST_PRECINCT'])[['ARREST_KEY']].count().reset_index().rename(columns={'ARREST_KEY':'ARRESTOS'})

In [131]:
precinto_peligroso=precinto_peligroso[precinto_peligroso['ARREST_PRECINCT']==14]

In [132]:
# Se suaviza la serie de tiempo antes de graficar
precinto_peligroso2 = precinto_peligroso.copy()
precinto_peligroso2.iloc[:,1:]=precinto_peligroso2.rolling(window=20).mean().fillna(0)

In [133]:
#Se crea la gráfica
fig = px.line(precinto_peligroso2, x='FECHA', y ="ARRESTOS", title = '<b>Evolución del precinto 14<b>')

fig.update_layout(
    xaxis_title = 'Año',
    yaxis_title = 'Arrestos',
    template = 'simple_white',
    title_x = 0.5)
fig.show()

**Interpretación:** A través del tiempo, el promedio de arrestos por año del distrito 14 ha disminuido a partir del año 2015 aproximadamente, ademas se encuentra en su punto mas bajo entre el año 2020 y 2021.

<p><a name="s3"></a></p>

# **Pregunta 10**



**¿Cuál es el género que más se presenta en los perpetradores por nivel de ofensa
(LAW_CAT_CD)?**

In [134]:
Ofgenero=Seriefechacrimenes.groupby(['PERP_SEX','LAW_CAT_CD'])[['ARREST_KEY']].count().reset_index().sort_values('ARREST_KEY', ascending=False)
Ofgenero=Ofgenero[Ofgenero['LAW_CAT_CD']!='Desconocido']
Ofgenero

PERP_SEX LAW_CAT_CD  ARREST_KEY
8        M          M     2897027
6        M          F     1275864
3        F          M      642988
9        M          V      273268
1        F          F      239395
7        M          I       24766
4        F          V       20139
2        F          I        1844

In [135]:
#Se crea un diccionario para reemplazar los nombres de los niveles de ofensa:

dicc={'M':'Delito menor','F':'Delito grave','V':'Violacion','I':'Informatico'}

Ofgenero['LAW_CAT_CD']=Ofgenero['LAW_CAT_CD'].replace(dicc)

In [137]:
# Grafica
df = px.data.tips()
fig = px.sunburst(Ofgenero, path=['LAW_CAT_CD','PERP_SEX'], values='ARREST_KEY',color='ARREST_KEY')
fig.show()

**Interpretación:** teniendo en cuenta que el delito menor es el que más de presenta en los arrestos, se identifica que el genero masculino es el que predomina respecto al femenino. igualmente sucede con el delito grave, violacion e informatica, aunque este ultimo se presenta en menor medida.

<p><a name="s3"></a></p>

# **Pregunta 11**



**¿En qué meses ocurren más arrestos? ¿Qué recomendaría hacer en estos meses?**

In [138]:
#Se agrupa por fecha el total de arrestos
Arrestos=Seriefechacrimenes.groupby(['FECHA'])[['ARREST_KEY']].count().reset_index()

In [139]:
# Se decide filtrar excluyendo el año 2022 ya que este no cuenta con los meses de julio,agosto,septiembre,octubre,noviembre y diciembre.
Arrestoshistoricos=Arrestos[Arrestos['FECHA'].dt.year != 2022]

In [140]:
# Se saca el mes
Arrestoshistoricos=Arrestoshistoricos.resample('1M',on ='FECHA').sum().reset_index()
Arrestoshistoricos['FECHA']=Arrestoshistoricos['FECHA'].dt.month

In [141]:
# Se crea un diccionario para organizar los meses
meses={1:'Enero',2:'Febrero',3:'Marzo',4:'Abril',5:'Mayo',6:'Junio',7:'Julio',8:'Agosto',9:'Septiembre',10:'Octubre',11:'Noviembre',12:'Diciembre'}

In [142]:
Arrestoshistoricos['MES']=Arrestoshistoricos['FECHA'].replace(meses)

In [143]:
Arrestoshistoricos=Arrestoshistoricos.groupby(['MES','FECHA'])[['ARREST_KEY']].sum().reset_index().sort_values('FECHA',ascending=True)
Arrestoshistoricos.head()

MES  FECHA  ARREST_KEY
3    Enero      1      468027
4  Febrero      2      435720
7    Marzo      3      482038
0    Abril      4      452894
8     Mayo      5      471045

In [146]:
# crear gráfica
fig = px.bar(Arrestoshistoricos, x='MES', y='ARREST_KEY', title ='<b>Arrestos por mes<b>', color_discrete_sequence=px.colors.qualitative.Set3, text_auto='.2s')

# agregar detalles a la gráfica
fig.update_layout(
    xaxis_title = 'Mes',
    yaxis_title = 'Promedio cantidad de arrestos',
    template = 'simple_white',
    title_x = 0.5)
fig.update_traces(textfont_size=12, textangle=0, textposition="outside", cliponaxis=False)
fig.show()

**Interpretación:** A pesar de que la media de arrestos en cada mes es similar y esta en una media de 450 mil arrestos aproximadamente, el mes con mayor cantidad de arrestos es marzo y el de menor cantidad es diciembre, se recomienda tener presencia de las autoridades en todos los meses ya que como se refleja en los resultados, la cantidad de arrestos se podría decir que no varía tan significativamente entre meses.

<p><a name="s3"></a></p>

# **Pregunta 12**


**¿Cuál es la raza y rangos de edad que se presentan en los perpetradores?**

In [147]:
Raza=Seriefechacrimenes.groupby(['AGE_GROUP','PERP_RACE'])[['ARREST_KEY']].count().sort_values('ARREST_KEY',ascending=False).reset_index()

In [148]:
#Se excluye la edad Desconocida
Raza=Raza[Raza['AGE_GROUP']!='Desconocido']

In [149]:
# Se crea un diccionario para organizar los grupos de edades
edad={1:'<18',2:'18-24',3:'25-44',4:'45-64',5:'65+'}

In [150]:
Raza['AGE_GROUP']=Raza['AGE_GROUP'].replace(edad)

In [151]:
Raza=Raza.sort_values('ARREST_KEY',ascending=False)
Raza.head()

AGE_GROUP       PERP_RACE  ARREST_KEY
0     25-44           BLACK     1167275
1     25-44  WHITE HISPANIC      695969
2     18-24           BLACK      676135
3     45-64           BLACK      525054
4     18-24  WHITE HISPANIC      374015

In [156]:
# crear gráfica
fig = px.bar(Raza, x='AGE_GROUP', y='ARREST_KEY', color ='PERP_RACE', barmode ='group',category_orders={"AGE_GROUP": ["<18", "18-24", "25-44", "45-64","65+"]},
             title ='<b>Las raza y rangos de edad que se presentan en los perpetradores<b>',color_discrete_sequence=px.colors.qualitative.Set3, text_auto='.2s')

# agregar detalles a la gráfica
fig.update_layout(
    xaxis_title = 'Grupos de edad',
    yaxis_title = 'Cantidad de crimenes',
    legend_title = 'Tipos de razas',
    template = 'plotly_white',
    title_x = 0.5)
fig.update_traces(textfont_size=8, textangle=0, textposition="outside", cliponaxis=False)
fig.show()

**Interpretación:** La raza que más se ve presente en los arrestos es la negra sin importar la edad,ademas se observa que el rango que mas presenta tendencias criminales esta entre la edad de 25 a 44 años, por ejemplo la raza negra con más de 1 millón de arrestos.Igualmente se destaca la presencia de la raza blanca hispana en lugar en todos los rangos de edad. El rango de edad donde hay una menor cantidad de arrestos es 65+.

<p><a name="s3"></a></p>

# **Conclusiones**



- El barrio más peligroso por población y área en Nueva York es Manhattan, seguido de el Bronx, puesto que en ellos se presentaron la mayor cantidad de personas arrestadas por diferentes delitos.
- El delito más frecuente es el drogas peligrosas, específicamente hablando de posesión, consumo y venta de drogas,como la marihuana.
- El día de la semana en el cual se presentan más arrestos es entre martes y miércoles y en el que menos se presentan son los domingos y lunes. Según los datos históricos y actuales.
- El mes en el que ocurren más arrestos es en marzo y en dicimbre es donde ocurren menos arrestos, sin embargo, los demás meses tienen un promedio de arresto similar.
- La raza negra, entre los 25 a 44 años de edad son más propensos a estar involucrados en estos arrestos.
- El precinto más peligroso respecto a los arrestos, es el #14 y luego está el #75, dependiendo del mes que se analice.
-Se encontraron muchos datos nulos y atípicos mediante la exploración de la base de datos, en su mayoría correspondia a la escritura de códigos, barrios y rangos de edades de las personas, por lo cual se recomienda que la policia tenga una mejor capacitación y manejo de los datos que ingresan para evitar que se pierda tanta información que puede ser útil en estudios futuros.